### ISSUE TITLE
Is Cost actually a Layer?

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#1311]
It seems that it is.

Consider the following network for image classification:

```
image -> conv -> softmax
```

If we are going to train it by minimizing the squared error, we can add

```
image -> conv -> softmax \
                          -> squared_error
                 label --/
```

So it seems that the "model" should contain only the part from image to softmax, and at training time we add label and squared_error.

**Comment by jacquesqiao** [comment]
Yes, I agree with you.
According to the discussion of in the last few days, Cost is not a layer but a independent concept, the topology of neural network only contains image -> conv -> softmax, Cost will be used in train and test.

**Comment by wangkuiyi** [comment]
Hi @jacquesqiao ,  @helinwang and I learned something more from @emailweixu this afternoon that **it is not flexible enough to express the *to-be-trained network* by a composition of *model* and *cost* **.

**Comment by wangkuiyi** [comment]
By this conclusion, I am closing this issue and open another issue for the discussion of the concept *network*.
